### ISSUE TITLE
feature request: crafted-package-install-package with guix

### ISSUE THREAD

**Issue by jgarte** [SystemCrafters/crafted-emacs#209]
Hi, are there any plans to use guix as another backend for `crafted-package-install-package` besides straight?

https://github.com/alezost/guix.el/search?l=Emacs+Lisp&q=install

**Comment by oatmealm** [comment]
Would be greatly appreciated. At the moment the package macro seems to ignore Guix emacs packages installed into the user profile.

**Comment by jeffbowman** [comment]
It might be nice to have a Guix backend for packages. If you have a PR that would be great! I don't personally use Guix, perhaps @daviwil  would be able to help with this one.  There is `crafted-ensure-package` but I don't think it is very useful other than to tell you a package is not installed by `guix`.  Definitely needs some work here. 

I can start looking into it, but someone who uses `guix` to configure Emacs would be better.

**Comment by oatmealm** [comment]
@jeffbowman I think at first check if availabe before installing from melpa etc. As far as I undestand the only check is in `crafted-package-install-package` which simply calls `package-installed-p` to check if package is available via `package.el` only, I assume, and not in `EMACSLOADPATH` ....

**Comment by oatmealm** [comment]
On the same issue, I think it might make sense to let a user to have their own manifest for the packages needes, so maybe `Crafted` should ship with a default manifest files (which you already have for `guix shell` I assume)... maybe it could include all the packages from the enabled/available modules?

**Comment by abarocio80** [comment]
For what I have learned, to have `guix` configured is the best option, may it be with `guix install`, a manifest (as proposed by @oatmealm), system configuration or home configuration.

The option to include a manifest file is a good idea, since it could also serve as a starting point for the users to write their own. It could be commented for most packages and let the users to uncomment by them selfs.

I do use guix, and I have been looking for this problem almost all week (I actually thought on writing an emacs package for this), but I learned that the best way to go is outside of emacs.  Not even with `emacs-guix` could you install the packages and leave those usable in the same session; you need to restart your emacs session.

**Comment by jgarte** [comment]
I think someone may be working on the issue that @abarocio80 detailed in Guix. It's also an issue for loading common lisp libraries installed with Guix.

**Comment by abarocio80** [comment]
Well, I have making some inqueries about the guix issue, and **I'm still convinced that the manifest is the best way to go**. But I still have found other things that I deem interesting for consideration, precisely for compatibility with `crafted-package-install-package`:

- A possible implementation for an emacs command to install a package with guix:
  1. I found a way to add packages installed by `guix` to the `load-path` without the need to restart, which is as simple as load the file `share/emacs/site-lisp/subdirs.el` from the `guix-profile` directory (`~/.guix-profile` in my system), but I'm not sure it is a good idea. Also, some packages are installed in the `share/emacs/site-lisp` directory (e.g. `mu4e`, wich is included in the system package `mu`), so that directory needs also to be added in the `load-path`. I'm not sure if that file is automatically loaded on startup. (@jgarte mentioned this)
  2. I have found that the most concise way to detect if a package is installed is with the function `locate-library`.
  3. I'm struggling to install packages with emacs-lisp (like with an emacs command/function) in a concise manner. I've tried with `shell-command`, `call-process` and `start-process`, but I don't like the output; further more, I still need to find a way to decide if a package was successfully installed or not.
- Not all the packages have just the suffix `emacs-` attached to their elpa name (e.g. `emacs-straight-el`).
- Not all the elpa packages are available in guix (via `guix` channel or the `nonguix` channel), but it's almost trivial to import a package definition with `guix import` that could be used for installing packages from elpa with `guix package -f` (and store those files in a subdirectory of `crafted-config-path`).
- In the same directory, the users could add their own package-description files to install packages that are not available in guix nor in elpa (like `ement.el` from alphapapa).
- I also have been thinking on making possible to generate a manifest directly with the theoretical command `crafted-guix-package-install` (which would be the backend to `crafted-package-install-package` if `crafted-package-system` is set to `'guix`), so the users could have a stateless description of their packages ready to be installed in other systems/environments.

There are other ideas I have had, but these are the most straight forward ones. Sadly, right now I'm struggling with other issues in my laptop, so I won't be able to flesh-out anything in a short period; if any one want's to pickup these ideas (all or any), feel free.
